### DISCUSSION TITLE
[3/11] Basics of Deep Learning

### DISCUSSION THREAD

**Discussion by tjdtnsu** [Pseudo-Lab/EfficientML#3] [General]
1. 전체적인 강의 내용
2. 이번에 특별히 알게 된 것 하나만 자세히 소개해보기 
3. 강의 중 이해가 더 필요했던 것
4. 이 강의의 중요 포인트 

위에 껀 예시. 강의 듣고 알게 된 내용, 느낀 점들 등등 적고 싶은 것들 많이많이 써보기

**Comment by tjdtnsu** [comment]
복습하는 파트여서인지 거의 다 알고 있었다고 생각했지만, 이번에 강의를 들으면서 다시 한 번 개념을 이해하는 느낌으로 들었음. 
잘 몰랐던 것을 새로 알게 된 것도 있었음.

전체적으로는 FC, Conv, Normalization을 먼저 다루고, Efficiency 관련해서 어떤 점을 고려해야 하는지 배우는 시간이 있었음.
Conv receptive Field L(k-1)+1 공식 처음 보았고, stride/pooling이 있을 때 어떻게 계산해야 할까 생각해보게 되었음 (PatchGAN)
이전에도 헷갈렸고, 이번에 들으면서 그나마 감 잡은 Normalization은 어느 축에서 되는건가.. 유지되는가
Activation이 이런게 있다.. 정도보다는 효율성과 gradient 차원에서 들은 점이 흥미가 꽤 있었음. Hard Swish는 거의 들어보지도 못했는데, quantization과 시간 복잡도 측면에서 왜 좋은 연산인지 알게 되었음.
Latency와 Throughput은 반대의 개념인 줄 알았는데, 아닐 수도 있다!라는 것도 알게 됨 (병렬로 하면 되군..)
이번에 거의 처음 들은 개념은 data movement of activation인데, 이는 메모리 대역폭과 activation size와 관계가 있군! 이 부분 이해를 잘 못해서 더 찾아봐야 함. 알게 된 건 메모리 옮기는 데 에너지가 많이 들게 된다는 점. 옮기는데 시간은 적게 드나?
Peak Activation이 메모리쪽과 관련이 있는 듯 하다. 왜 input activation, output activation 두개를 더해야 할까? 저장 개념으로 봐야 할 듯
MAC과 FLOP이 다른 개념인지도 처음 알게 되었음. MAC은 a<-a+b*c, FLOP은 더하기 곱하기 각각 한 연산으로 치는 듯. 그래서 FLOP이 MAC*2인가? 다른 경우도 있나?
OP는 float가 아닌 연산에도 활용할 수 있다는데 어디에 쓸까? 당연히 quantization이겠지?

**Comment by namgyu-youn** [comment]
### 전체적인 강의 내용
전체적으로 개념에 대해서 복습하는 강의였습니다. FC, Convolution, Pooling, Normalization처럼 Pytorch에서 사용하던 내용들부터 시작해서, CV & NLP를 지배중인 Transformer까지 되돌아보는 것으로 시작합니다.

그 다음은 경량화에 사용되는 efficiency metrics에 대해서 소개합니다. Parameters, Model size (memory)처럼 기존에 알던 개념 뿐만 아니라, latency & energy 측면에서도 소개하면서 TinyML의 중요성에 대해서 강조합니다. 특히 latency는 real-time service에서 중요한 개념이므로 흥미롭게 다가왔습니다. 

### 이번에 특별히 알게 된 것 하나만 자세히 소개해보기
이전에 vanilla ViT를 공부하면서 경량화에 대한 관심이 처음 생겼던 것 같습니다. ViT의 특징은 전체 sequence를 활용해 relationship을 파악하기 때문에 성능 향상이 있었지만, 전체 sequence를 한 번에 학습해야하기 때문에 cost 한계가 있다는 점입니다. (첫 학습 비용 어마무시함)

아직 관련 지식이 부족하며 논문 리뷰를 해본 것은 아니지만, 강의에서 잠깐 등장하는 EfficientViT를 보면서 반가웠던 것 같습니다. 빨리 이론 지식을 학습한 뒤에, 논문 리뷰를 해보고 싶다는 생각을 하게 되었습니다.

### 강의 중 이해가 더 필요했던 것
MACs, FLOPs, OPS같은 efficiency metrics에 대한 이해가 더 필요한 것 같습니다. 이전에 모델을 학습시킬 때는 OOM 발생하면 아 터졌구나.. 더 가볍게 돌려야겠네? 라는 막연한 결론밖에 내리지 못했던 기억이 있습니다. 그리고 최근에 하고 있는 생각이지만, CUDA를 공부해야할 것 같습니다...

### 이 강의의 중요 포인트
개인적으로 Transformer based model에서 경량화가 더 궁금했는데, 강의 앞부분은 CNN 기반 모델들에 대해서만 이야기하는 느낌 같아서 아쉬웠습니다. 12강부터 Transformer 이야기가 시작되며 16강부터 정말 궁금한 내용이 등장할 것 같은데, 이번 기수 안에 끝낼 수 있을지도 걱정되네요. 시간 있을 때마다 빠르게 예습해야할 것 같습니다.

**Comment by Saebitna** [comment]
### 강의 요약 
- 기본적인 용어, neural network layers에 대한 소개
- Efficiency Metrics에 대한 소개 

### 새로 배운 점
- Grouped Convolution Layer & Depthwise Convolution Layer : AlexNet, RexNeXt, MobileNet 등에 적용해서 연산량을 감소시킴. 그룹 별로 독립적으로 convolution을 수행한다는 아이디어가 좋음 
- Normalization Layer에서 주로 사용하는 batch norm, layer norm 외 instance norm, group norm 을 배움. 어디에 적용할 수 있을까 궁금함. 
- Activation Function : ReLU6, Swish, Hard Swish를 배움. 성능과 연산 최적화를 고려했을 때 어떤 activation functon이 프로젝트에 따라 적합한지 실험해 볼 가치가 있어보임. 
- Efficiency Metrics : 대부분 처음 배운 내용으로 개인적으로 가장 의미가 있었음. Latency vs Throughput에서 두 개념이 상반된 개념이 아니며, 병렬처리의 중요성과 Latency가 무조건 낮다고 좋은 게 아님을 느낌. 
- 기타: 66p의 그래프에서 activation function의 수 와 파라미터의 수를 어떻게 조정해야 효율적으로 메모리를 사용할 수 있을까 하는 개인적인 호기심이 생김. 
- MAC 개념은 아직 확실히 이해가 안되어서 복습이 필요해 보임 

### 느낀 점
- 딥러닝 모델 학습 시 어디서 오래 걸리고, 병목현상이 생기는지, 어떻게 연산량 등을 측정하는지에 대해 배움. 모델을 학습할 때 주로 학습이 오래 걸리거나 메모리 부족으로 학습이 중단될 때, 인풋 데이터 크기, layer depth를 줄이거나 메모리 할당량을 늘려서 문제를 해결하려고 했는데 이제 engineer 팀과 일할 때 큰 도움이 될 것으로 보임

**Comment by surisurikim** [comment]
### 1. Review the terminology of neural networks
- Neuron, Synapses, Activation, Feature, Weight, Parameter, etc

### 2. Review popular building blocks in a neural network
- Fully-Connected, Convolution, Grouped Convolution, Depthwise Convolution
- Pooling, Normalization, Transformer

### 3. Introduce popular efficiency metrics for neural networks
- Parameters, Model Size, Peak #Activations, MAC, FLOP, FLOPS, OP, OPS, Latency, Throughput
    - **새롭게 알게 된 내용**
        - Latency와 Throughput이라는 metric
        - Model size를 파라미터 수가 아닌 용량 그 자체로 표기하는 방법
        - Activation에 대한 논의
            - **Inference(추론) 시 가장 중요한 메모리 병목은 Activation 크기이며,** 모델이 실행될 때, **가장 메모리를 많이 차지하는 순간(Peak Activation)이 전체 메모리 제한을 결정할 수 있다.**
            - 특정 레이어(예: 네트워크 초반부)의 Activation이 너무 크면, **전체 모델이 동작 불가능할 수도 있음.**
        - MAC이라는 표현 방식
    - **궁금한 점**
        **- 강의 전반적으로 아는 내용 리마인드였지만 마지막에 MAC이랑 FLOP은 좀 생소합니다. 이 부분 실무에서 어떤 식으로 언급되고 사용되는지? 경험 있으신 분들의 이야기 궁금합니다.**

**Comment by hoodong** [comment]
Basic이 꽤 어렵네요!!!
2시간 투자했고 그래도 시간이 부족해서 트랜스포머는 자체 스킵했어요. 
(이직해서 신입이라 업무량이 많아서.... 더 분발하겠습니다!)
MAC와 FLOP 등 계산량 용어는 알던 내용이라 그나마 이해되네요.
활성화함수가 메모리 병목에 큰 영향을 준다는 사실이 새롭웠고,
하지만 그 매커니즘이 잘 이해되지 않네요! 
활성화 함수가 어떻게 메모리 병목에 영향을 주는지 모르겠어요...

**Comment by sw930718** [comment]
### 전체적인 강의 내용
- neural network 용어 복습
    - Neuron, Synapses, Activation, Feature, Weight, Parameter, etc
- neural network 내 유명한 building blocks 복습
    - Fully-Connected, Convolution, Grouped Convolution, Depthwise Convolution
    - Pooling, Normalization, Transformer
- neural network에 대한 유명한 성과지표 소개
    - #Parameters, Model Size, Peak #Activations, MAC, FLOP, FLOPS, OP, OPS, Latency, Throughput
    
### 강의 중 이해가 더 필요했던 것
Depthwise Convolution

### 이 강의의 중요 포인트
neural network와 관련된 용어 정리와, 각 구조에 대한 이해, 다양한 성과지표에 대한 이해와 trade-off 관계들 

### 느낀점
배웠던 내용중에서도 제대로 이해 안하고 넘어갔던게 많았던거 같아서, 이번 기회에 정리할수있어서 좋았습니다.
